### PR TITLE
Module Machine Learning: reduction NN

### PR DESCRIPTION
**Pull request by LauraMeneghetti**

### CODE REVIEW

File: smithers/ml/fnn.py [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+                      (epoch + 1, i + 1, running_loss / 50))
+                if i == inputs_net.size()[0] // batch_size - 1:
+                    final_loss.append(running_loss / 50)
+                running_loss = 0.0
+
+            _, predicted = torch.max(outputs.data, 1)
+            labels = torch.LongTensor(real_out[i * batch_size:(i + 1) *
+                                               batch_size])
+            total += labels.size(0)
+            correct += (predicted == labels).sum().item()
```
Comment (ndem0) [inline smithers/ml/fnn.py:107]: I would keep the FNN customizable by user (right now the architecture is hardcoded). Maybe we can use the model from [PINA](https://github.com/mathLab/PINA/blob/main/pina/model/feed_forward.py)?

File: smithers/ml/netadapter.py [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+            It iterates over the given dataset, obtained combining a
+            dataset(images and labels) and a sampler.
+        :param int n _class: number of classes that composes the dataset
+        :return: reduced net
+        :rtype: nn.Module
+        '''
+        input_type = train_dataset.__getitem__(0)[0].dtype
+        possible_cut_idx = PossibleCutIdx(input_network)
+        cut_idxlayer = possible_cut_idx[self.cutoff_idx]
+        print('ff')
```
Comment (ndem0) [inline smithers/ml/netadapter.py:221]: remove print

File: smithers/ml/netadapter.py [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+    def __init__(self, cutoff_idx, red_dim, red_method, inout_method):
+        '''
+        :param int cutoff_idx: value that identifies the cut-off layer
+        :param int red_dim: dimension of the reduced space onto which we
+            project the high-dimensional vectors
+        :param str red_method: string that identifies the reduced method to
+            use, e.g. 'AS', 'POD'
+        :param str inout_method: string the represents the technique to use for
+            the identification of the input-output map, e.g. 'PCE', 'ANN'
+        '''
```
Comment (ndem0) [inline smithers/ml/netadapter.py:33]: I would add an example section with the MINIMAL but working example

File: smithers/ml/fnn.py (hunk 1 above)
Comment (LauraMeneghetti) [reply smithers/ml/fnn.py:107]: Ok. Should I import that class from PINA?
I think I have only to check that everything is working for the CIFAR dataset, where I have placed in the forward `x = x.view(x.size(0), -1)` in order to have everything working correctly. 
Then the function training_fnn should be removed and placed in the utils.py file or has already been implemented in PINA?

Comment (ndem0) [reply smithers/ml/fnn.py:107]: Yes, for now I would just import directly from PINA (in future we can evaluate other possibilities)

**Comment by ndem0** [comment]
Dear @LauraMeneghetti @szanin, the total amount of changed files is beyond 600, which I think is too many. The dataset should not be uploaded, just put the script to create/download it (Idk if it's already in the repo)

File: smithers/ml/AHOSVD.py [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+        :return: the updated list of averages
+        :rtype: list
+        """
+        matrices_list = []
+        if index == 0:
+            return new_list
+        elif index > 0:
+            for i in range(len(current_list)):
+                matrices_list.append((index / (index + 1)) * current_list[i] + (1/(index + 1)) * new_list[i])
+            return matrices_list
```
Comment (ndem0) [inline smithers/ml/AHOSVD.py:44]: raise ValueError in case index < 0

File: smithers/ml/AHOSVD.py [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+        """
+        Computes the partial HOSVD from a restricted sample of the snapshots tensor
+
+        :param torch.Tensor batch_from_tensor: the batch given
+        :return: list of U matrices coming from the modal SVDs
+        :rtype: list[torch.Tensor]
+        """
+        HOSVD = hosvd(batch_from_tensor.shape)
+        #HOSVD.fit(batch_from_tensor)
+        #return HOSVD.modes_matrices[1:]
```
Comment (ndem0) [inline smithers/ml/AHOSVD.py:56]: remove comments

**Comment by ndem0** [comment]
@LauraMeneghetti Now the tests are passing (except for one issue due to scipy in py38, not related to this PR), but the folder organization is quite confusing, even to me that I know the math behind it. Do we really need 20 new files? In `models`, only few files actually implement models (fnn, vgg, rednet).
I would clean the files first of all: can we have the folders `model`, `layer`, `loss` with the related files inside them? Not all the files should go there anywat!
